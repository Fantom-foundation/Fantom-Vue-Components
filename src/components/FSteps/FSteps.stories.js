import FSteps from './FSteps.vue';

export default {
    title: 'FSteps',
    component: FSteps,
};

export const Default = () => ({
    components: { FSteps },
    template: `
        <div>
            <f-steps :count="3" />
        </div>
    `,
});

export const Active = () => ({
    components: { FSteps },
    template: `
        <div>
            <f-steps :count="4" :active="2" />
            <f-steps :count="4" :active="20" />
            <f-steps :count="4" :active="0" />
        </div>
    `,
});

export const WithLabels = () => ({
    components: { FSteps },
    template: `
        <div>
            <f-steps :labels="labels" />
            <f-steps :labels="labels" :active="2" />
            <f-steps :labels="labels" :active="3" />
        </div>
    `,
    data() {
        return {
            labels: ['Step 1', 'Step 2', 'Step 3'],
        };
    },
});
