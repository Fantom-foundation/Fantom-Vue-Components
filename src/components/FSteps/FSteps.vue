<template>
    <div class="fsteps" :class="{ 'labels-on': labels.length > 0 }">
        <div class="fsteps_cont">
            <div
                v-for="(step, index) in steps"
                :key="step.id"
                class="fsteps_step"
                :class="{ 'fsteps_step-active': index < cActive }"
            >
                <span class="fsteps_label" :class="{ 'fsteps_label-active': index < cActive }">
                    {{ step.label }}
                </span>
            </div>
        </div>
    </div>
</template>

<script>
import { getUniqueId } from '../../utils';

export default {
    name: 'FSteps',

    props: {
        /** Count of steps */
        count: {
            type: Number,
            default: 0,
        },
        /** Active step (`<1, count>`) */
        active: {
            type: Number,
            default: 1,
        },
        /** Step labels. */
        labels: {
            type: Array,
            default() {
                return [];
            },
        },
    },

    data() {
        return {
            /** */
            steps: [],
        };
    },

    computed: {
        /**
         * @return {number}
         */
        cActive() {
            const { stepsCount } = this;

            if (this.active < 1) {
                return 1;
            } else if (this.active > stepsCount) {
                return stepsCount;
            } else {
                return this.active;
            }
        },

        /**
         * @return {number}
         */
        stepsCount() {
            return this.labels.length > 0 ? this.labels.length : this.count;
        },
    },

    created() {
        this.prepareSteps();
    },

    methods: {
        prepareSteps() {
            const { stepsCount } = this;
            const { labels } = this;
            this.steps = [];

            for (let i = 0; i < stepsCount; i++) {
                this.steps.push({
                    id: getUniqueId(),
                    active: false,
                    label: labels.length > 0 ? labels[i] : '',
                });
            }
        },
    },
};
</script>

<style lang="scss">
@use 'style';
</style>
