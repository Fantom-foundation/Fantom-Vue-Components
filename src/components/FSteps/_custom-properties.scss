.fsteps {
    --fsteps-padding: 8px;

    /* main colors */
    --fsteps-primary-color: var(--f-color-primary-5, #1a6aff);
    --fsteps-secondary-color: #e6e6e6;

    /* horizontal line */
    --fsteps-hor-line-height: 1px;
    --fsteps-hor-line-color: var(--fsteps-secondary-color);
    /* active horizontal line */
    --fsteps-active-hor-line-background: var(--fsteps-primary-color);

    /* steps container */
    --fsteps-cont-padding: 0;
    --fsteps-cont-background: linear-gradient(var(--fsteps-hor-line-color), var(--fsteps-hor-line-color)) no-repeat
        center/100% var(--fsteps-hor-line-height); /* horizontal line */

    /* step dot */
    --fsteps-step-dot-width: 16px;
    --fsteps-step-dot-height: 16px;
    --fsteps-step-dot-border: var(--fsteps-hor-line-height) solid var(--fsteps-hor-line-color);
    --fsteps-step-dot-border-radius: 50%;
    --fsteps-step-dot-content: '';
    --fsteps-step-dot-background: var(--fsteps-secondary-color);
    /* active step dot */
    --fsteps-step-dot-active-background: var(--fsteps-primary-color);
    --fsteps-step-dot-active-border: var(--fsteps-hor-line-height) solid var(--fsteps-step-dot-active-background);

    /* label */
    --fsteps-label-color: inherit;
    --fsteps-label-width: 60px;
    --fsteps-label-padding: 4px 0 0 0;
    /* active label */
    --fsteps-label-active-color: var(--fsteps-primary-color);
}
