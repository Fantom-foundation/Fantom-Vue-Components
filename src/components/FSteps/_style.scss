@use 'custom-properties';

.fsteps {
    position: relative;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    padding: var(--fsteps-padding);
    /*background: linear-gradient(#eee, #eee) no-repeat center/100% 1px;*/

    &.labels-on {
        padding-bottom: 32px;
    }

    &_cont {
        width: 100%;
        display: flex;
        padding: var(--fsteps-cont-padding);
        background: var(--fsteps-cont-background);
    }

    &_step {
        position: relative;
        flex: 1;
        display: flex;
        justify-content: flex-end;
        width: 100%;

        &:first-child {
            flex: 0;
        }

        &::before {
            width: var(--fsteps-step-dot-width);
            height: var(--fsteps-step-dot-height);
            background: var(--fsteps-step-dot-background);
            border: var(--fsteps-step-dot-border);
            border-radius: var(--fsteps-step-dot-border-radius);
            content: var(--fsteps-step-dot-content);
            display: block;
        }

        &-active {
            &::before {
                position: relative;
                background: var(--fsteps-step-dot-active-background);
                border: var(--fsteps-step-dot-active-border);
                z-index: 1;
            }

            &::after {
                position: absolute;
                top: 50%;
                transform: translateY(-50%);
                right: 0;
                width: 100%;
                height: var(--fsteps-hor-line-height);
                background: var(--fsteps-active-hor-line-background);
                content: '';
                z-index: 0;
            }
        }
    }

    &_label {
        position: absolute;
        color: var(--fsteps-label-color);
        width: var(--fsteps-label-width);
        padding: var(--fsteps-label-padding);
        text-align: center;
        transform: translateX(calc(25% + 0.5 * var(--fsteps-step-dot-width))) translateY(50%);

        &-active {
            color: var(--fsteps-label-active-color);
        }
    }
}

html[dir=rtl], body[dir=rtl] {
    .fsteps_label {
        transform: translateX(calc(-50% + 0.5 * var(--fsteps-step-dot-width))) translateY(50%);
    }
}
