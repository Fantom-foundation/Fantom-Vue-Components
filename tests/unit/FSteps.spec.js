import { mount } from '@vue/test-utils';
import FSteps from '@/components/FSteps/FSteps.vue';

let wrapper = null;

function destroyWrapper(wrapper) {
    if (wrapper) {
        wrapper.destroy();
        wrapper = null;
    }
}

function createWrapper({ propsData = {}, slots = {} } = {}) {
    return mount(FSteps, {
        propsData,
        slots,
    });
}

afterEach(() => {
    destroyWrapper(wrapper);
});

describe('FSteps', () => {
    it('should render correctly', () => {
        wrapper = createWrapper();

        expect(wrapper.element).toMatchSnapshot();
    });

    it('should draw steps if `count` prop is given', async () => {
        wrapper = createWrapper({
            propsData: {
                count: 3,
            },
        });

        expect(wrapper.findAll('.fsteps_step').length).toBe(3);
    });

    it('should draw active steps if `active` prop is given', async () => {
        const activeClass = 'fsteps_step-active';

        wrapper = createWrapper({
            propsData: {
                count: 3,
                active: 2,
            },
        });

        const steps = wrapper.findAll('.fsteps_step');

        expect(steps.at(0).classes(activeClass)).toBeTruthy();
        expect(steps.at(1).classes(activeClass)).toBeTruthy();
        expect(steps.at(2).classes(activeClass)).toBeFalsy();
    });

    it('should draw labels if `labels` prop is given', async () => {
        wrapper = createWrapper({
            propsData: {
                count: 2,
                labels: ['step 1', 'step 2'],
            },
        });

        const labels = wrapper.findAll('.fsteps_label');

        expect(labels.at(0).text()).toBe('step 1');
        expect(labels.at(1).text()).toBe('step 2');
    });
});
