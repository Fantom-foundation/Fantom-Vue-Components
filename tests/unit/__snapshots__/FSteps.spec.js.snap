// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FSteps should render correctly 1`] = `
<div
  class="fsteps"
>
  <div
    class="fsteps_cont"
  />
</div>
`;
